src/testFixtures/groovy/grails/plugin/geb/ContainerGebSpec.groovy

```diff
@@ -44,38 +47,38 @@ import java.time.Duration
  * @author Mattias Reichel
  * @since 5.0.0
  */
-class ContainerGebSpec extends GebSpec {
+@CompileStatic
+abstract class ContainerGebSpec extends GebSpec implements ContainerAwareDownloadSupport {
 
+    private static final String DEFAULT_HOSTNAME_FROM_CONTAINER = 'host.testcontainers.internal'
+    private static final String DEFAULT_HOSTNAME_FROM_HOST = 'localhost'
     private static final String DEFAULT_PROTOCOL = 'http'
-    private static final String DEFAULT_HOSTNAME = 'host.testcontainers.internal'
+
+    private String hostNameFromContainer = DEFAULT_HOSTNAME_FROM_CONTAINER
 
     @Shared
     BrowserWebDriverContainer webDriverContainer
 
     @PackageScope
     void initialize() {
-        if (!webDriverContainer) {
-            if (!hasProperty('serverPort')) {
-                throw new IllegalStateException('Test class must be annotated with @Integration for serverPort to be injected')
-            }
-            webDriverContainer = new BrowserWebDriverContainer()
-            Testcontainers.exposeHostPorts(serverPort)
-            webDriverContainer.tap {
-                addExposedPort(serverPort)
-                withAccessToHost(true)
-                start()
-            }
-            if (hostName != DEFAULT_HOSTNAME) {
-                webDriverContainer.execInContainer('/bin/sh', '-c', "echo '$hostIp\t$hostName' | sudo tee -a /etc/hosts")
-            }
-            browser.driver = new RemoteWebDriver(webDriverContainer.seleniumAddress, new ChromeOptions())
-            browser.driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(30))
+        webDriverContainer = new BrowserWebDriverContainer()
+        Testcontainers.exposeHostPorts(port)
+        webDriverContainer.tap {
+            addExposedPort(this.port)
+            withAccessToHost(true)
+            start()
+        }
+        if (hostNameChanged) {
+            webDriverContainer.execInContainer('/bin/sh', '-c', "echo '$hostIp\t$hostName' | sudo tee -a /etc/hosts")
         }
+        WebDriver driver = new RemoteWebDriver(webDriverContainer.seleniumAddress, new ChromeOptions())
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(30))
+        browser.driver = driver
     }
 
     void setup() {
-        initialize()
-        baseUrl = "$protocol://$hostName:$serverPort"
+        if (notInitialized) initialize()
```

Comment:
Personally I prefer always to use curly brackets, even if there is just one line to execute in an if... Just a matter of style